extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/ImportFrameHandler.scala

```diff
@@ -41,28 +42,83 @@ class ImportFrameHandler extends Handler {
     val collectDomainsTask = new CollectCategoricalDomainsTask(frameKey)
     collectDomainsTask.doAllNodes()
     val stringDomains = collectDomainsTask.getDomains()
-    val domains = expandDomains(stringDomains, columnTypes)
+    val isEmpty = stringDomains == null
+    val columnTypesAwareOfEmptyFrames = if (isEmpty) {
+      columnTypes.map(columnType => if (columnType == Vec.T_CAT) Vec.T_STR else columnType)
+    } else {
+      columnTypes
+    }
+    val domains = expandDomains(stringDomains, columnTypesAwareOfEmptyFrames)
+    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypesAwareOfEmptyFrames, domains)
+    val frame = DKV.getGet[Frame](frameKey)
 
-    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypes, domains)
+    convertColumnsWithTooManyCategoricalLevelsToStringColumns(frame, columnTypesAwareOfEmptyFrames, stringDomains)
```

Comment:
this is a part I don't understand - the frame is already in DKV and has "too many" categoricals - should thee frame not be in DKV yet?

is there any locking in place, the ParseDataset locks the frame against reading

Reply:
The frame is in DKV from the beginning of the conversion. The initialize method puts the Frame into DKV (`ChunkUtils.initFrame`) as partial frame: 
```
    public static void initFrame(String keyName, String[] names) {
        Frame fr = new water.fvec.Frame(Key.<Frame>make(keyName));
        fr.preparePartialFrame(names);
        // Save it directly to DKV
        fr.update();
    }
```

Do you think that keeping the frame as partial and not finalizing it can avoid to read frame in a provisional state?

The columns with too many levels have null domains and during finalization are converted to `T_NUM` and then those columns are converted to `T_STR` with [ConvertCategoricalToStringColumnsTask](https://github.com/h2oai/sparkling-water/pull/2341/files/8ca5d9c1060fa02b7ce6d9edb6adf4a140bcc12b#diff-27d5892be9be15668ac386bf9b874876R27). 

I needed to finalize the frame first to get Vector instances which are [passed to ConvertCategoricalToStringColumnsTask](https://github.com/h2oai/sparkling-water/pull/2341/files/8ca5d9c1060fa02b7ce6d9edb6adf4a140bcc12b#diff-dcee268e9b2cd60b0fbd5684dc28114fR83).

Or is there any dedicated mechanism for locking frames for reading?

Reply:
After looking at the code mentioned by @mn-mikke, there is no reason why the frame should not be in DKV - at least to my understanding. H2O and presumably SW is a single-user environment. Therefore any accidental reads are not possible this way.

The original ParseDataset phase is finished once the frame is finalized. And there is no ParseDataset active during the `convertColumnsWithTooManyCategoricalLevelsToStringColumns` - or is there ? 

If not, then just removing the old vec (the `replace` method of a `Frame` is called) and replacing it with a new one seems to be safe operation. Chunk layout is adapted.

This is done sequentially.